x-pack/elastic-agent/docs/elastic-agent-dynamic-inputs.asciidoc

```diff
@@ -0,0 +1,837 @@
+of the provider in the context of the {agent}.
+
+For example, if a provider named **foo** provides
+`{"key1": "value1", "key2": "value2"}`, the key/value pairs are placed in
+`{"foo" : {"key1": "value1", "key2": "value2"}}`. To reference the keys, you
+would use `{{foo.key1}}` and `{{foo.key2}}`.
+
+=== Provider configuration
+
+The provider configuration is specified under the top-level `providers`
+key in the `elastic-agent.yml` configuration. By default, all registered
+providers are enabled. If a provider cannot connect, it produces no mappings.
+
+The following examples shows two providers, `local` and `local_dynamic`, that
+supply custom keys:
+
+[source,yaml]
+----
+providers:
+  local:
+    vars:
+      foo: bar
+  local_dynamic:
+    vars:
+      - item: key1
+      - item: key2
+----
+
+To explicitly disable a provider, set `enabled: false`. Because all providers
+are prefixed and have no name collisions, the name of the provider is the key in
+the configuration.
+
+[source,yaml]
+----
+providers:
+  docker:
+    enabled: false
+----
+
+{agent} supports two broad types of providers: <<context-providers,context>> and
+<<dynamic-providers,dynamic>>.
+
+[[context-providers]]
+=== Context providers
+
+Context providers provide the current context of the running {agent}, for
+example, agent information (id, version), host information (hostname, IP
+addresses), and environment information (environment variables).
+
+They can only provide a single key/value mapping. Think of them as singletons;
+an update of a key/value mapping will result in a re-evaluation of the entire
+configuration. These providers are normally very static, but that's not
+required. It is possible for a value to change resulting in re-evaluation.
+
+Context providers use ECS naming when possible to ensure that documentation and
+understanding across projects is the same.
+
+{agent} supports the following context providers:
+
+[[local-provider]]
+==== Local
+
+Provides custom keys to use as variables. For example:
+
+[source,yaml]
+----
+providers:
+  local:
+    vars:
+      foo: bar
+----
+
+[[agent-provider]]
+==== Agent provider
+
+Provides information about the {agent}. The available keys are:
+
+//PH: I think it might be easier for users if we show the keys they can use
+//in their config. Is this correct? 
+
+`agent.id`:: Current agent ID
+`agent.version`:: Current agent version information object
+`agent.version.version`:: Current agent version
+`agent.version.commit`:: Version commit
+`agent.version.build_time`:: Version build time
+`agent.version.snapshot`:: Version is snapshot build
+
+//PH: Is snapshot a boolean? Should we document the types here? 
+
+[[host-provider]]
+==== Host provider
+
+Provides information about the current host. The available keys are:
+
+`host.name`:: Host name
+`host.platform`:: Host platform
+`host.architecture`:: Host architecture
+`host.ip[]`:: Host IP addresses
+`host.mac[]`:: Host MAC addresses
```

Comment:
```
`host.name`:: Host name string
`host.platform`:: Host platform string
`host.architecture`:: Host architecture string
`host.ip[]`:: Host IP addresses []string
`host.mac[]`:: Host MAC addresses []string.
```

I don't think IP are detected automatically.

Reply:
@ph Are you saying that the user needs to do something extra to get the host IP?

Reply:
No, I've meant we don't have the type in ES and I think that ES will not auto-detect IPs. https://www.elastic.co/guide/en/elasticsearch/reference/current/ip.html